src/Servers/IIS/IIS/src/Core/IISHttpServer.cs

```diff
@@ -202,9 +202,9 @@ private static void OnRequestsDrained(IntPtr serverContext)
             {
                 server = (IISHttpServer)GCHandle.FromIntPtr(serverContext).Target;
 
-                server._nativeApplication.StopCallsIntoManaged();
-                server._shutdownSignal.TrySetResult(null);
-                server._cancellationTokenRegistration.Dispose();
+                server?._nativeApplication.StopCallsIntoManaged();
```

Comment:
This can be be null?

Reply:
yeah noticed when debugging after the process is shutting down when I was attached to the debugger.

Reply:
It looks like we complete the Task returned by StopAsync() as soon as CancellationToken fires. This means Hosting can dispose the IISHttpServer (which frees the GCHandle) before OnRequestsDrained gets called.

I bet it's not just OnRequestsDrained callback where the server reference can wind up being null. We also should be null checking the server reference in HandleRequest and HandleShutdown. I'm not sure if there's anywhere else this is necessary.

Reply:
This makes me uneasy. I'd like to better understand this before we start checking for null everywhere. Should we fix the draining logic?

Reply:
I feel uneasy with this too. Potentially we should setup a meeting to go over this logic?

Reply:
yah